docs/sysadmin/Configuration.md
# Configuration

To configure shipper, you need to set environment variables.

Because there are so many environment variables required for shipper, you may find that setting them all in a environment variable file and importing them before starting shipper is easier than manually setting them, one by one. To do this, create the files listed below, then run `set -a && source .env && set +a` to set all of the environment variables at once.

The files should look something like this:

`.sample.env`:

```
# Comments are supported, like this!
SHIPPER_SECRET_KEY="12345678"
SHIPPER_OTHER_CONFIG_OPTION=1 # for true
```


## `.env`

### Django internal
  - `SHIPPER_SECRET_KEY`
    - Django's secret key value, used internally for encryption, etc.
    - [Generate a random key](https://humberto.io/blog/tldr-generate-django-secret-key/) if you are starting a new shipper instance.
    - Required, no defaults provided.
  - `SHIPPER_DEBUG`:
    - Django's debug mode
    - `0` -> disabled (recommended for production)
    - `1` -> enabled (recommended for development)
    - Default: `0`
  - `SHIPPER_ALLOWED_HOSTS`
    - An array of allowed hosts in String format, with spaces delimiting each entry.
    - Example: if you're serving shipper at https://downloads.example.com, you would put `downloads.example.com` here.
    - Required, no defaults provided.
  - `SHIPPER_CSRF_TRUSTED_ORIGINS`
    - An array of trusted origins for CSRF in String format, with spaces delimiting each entry.
    - Should be similar to the `SHIPPER_ALLOWED_HOSTS` config key above, but **must** include the scheme (HTTP or HTTPS)
    - Example: if you're serving shipper at https://downloads.example.com, you would put `https://downloads.example.com` here.
    - Required, no defaults provided
  - `SHIPPER_CSRF_COOKIE_SECURE`
    - Internal Django security option.
    - `1` -> enabled (recommended for production)
    - `0` -> disabled
    - Default: `1`
  - `SHIPPER_SESSION_COOKIE_SECURE`
    - Internal Django security option.
    - `1` -> enabled (recommended for production)
    - `0` -> disabled
    - Default: `1`
  - `SHIPPER_SECURE_HSTS_SECONDS`
    - Internal Django security option to set up HSTS, a mechanism to only allow HTTPS traffic
    - `0` -> HSTS is disabled
    - Setting this key to any value other than 0 will enable HSTS for the specified duration (in seconds)
    - Warning: make sure you know what you are doing! Improper settings may block you from accessing your instance.
    - Default: `0`
  - `SHIPPER_TASK_TIME_LIMIT`
    - Controls the limit of build mirror tasks before they are killed
    - Setting is in minutes
    - Default: `60`

### Upload
  - `SHIPPER_UPLOAD_CHECKSUM`
    - Checksum type shipper expects when uploading via the chunked upload API
    - Supports the following options:
      - `md5`
      - `sha256`
    - Default: `sha256`

### Database
  - `SHIPPER_SQL_ENGINE`
    - Database type. Currently only supports the following options:
    - `django.db.backends.postgresql` -> PostgreSQL
    - Not specified -> SQLite
    - Default: SQLite
  - `SHIPPER_SQL_DATABASE`
    - Database name
    - Not specified -> SQLite
    - Default: SQLite
  - `SHIPPER_SQL_USER`
    - Database username
    - Default: `user`
  - `SHIPPER_SQL_PASSWORD`
    - Database password
    - Default: `password`
    - Warning: do NOT leave this variable unchanged in production!
  - `SHIPPER_SQL_HOST`
    - Database host
    - Default: `localhost`
  - `SHIPPER_SQL_PORT`
    - Database port
    - Default: `5432`

### Cache
  - `SHIPPER_CACHE_BACKEND`:
    - `django.core.cache.backends.memcached.PyMemcacheCache` -> Memory cache based on Memcached (recommended for production!)
    - `django.core.cache.backends.locmem.LocMemCache` -> Local, per-process memory cache (not recommended for production)
    - Default: `django.core.cache.backends.locmem.LocMemCache`
  - `SHIPPER_CACHE_LOCATION`:
    - Cache backend location.
    - If set to Memcached, specify a IP and port (`127.0.0.1:11211`) or a UNIX socket (`unix:/tmp/memcached.sock`)
    - If set to local, per-process memory cache, specify a name (`shipper-locmemcache`)
    - Default: `shipper-cache`

### Email
  - `SHIPPER_EMAIL_BACKEND`:
    - `django.core.mail.backends.console.EmailBackend` -> Local email (dummy option for debug use only!)
    - `django.core.mail.backends.smtp.EmailBackend` -> SMTP-based email (recommended for production!)
    - Default: `django.core.mail.backends.console.EmailBackend`
    - Warning: if set to `django.core.mail.backends.console.EmailBackend`, most of the keys below related to email will be ignored.
  - `SHIPPER_EMAIL_HOST`
    - Email (SMTP) host
    - Required if `SHIPPER_EMAIL_BACKEND` is set to `django.core.mail.backends.smtp.EmailBackend`
  - `SHIPPER_EMAIL_PORT`
    - Email (SMTP) host port
    - Required if `SHIPPER_EMAIL_BACKEND` is set to `django.core.mail.backends.smtp.EmailBackend`
  - `SHIPPER_EMAIL_HOST_USER`
    - Email username
    - Required if `SHIPPER_EMAIL_BACKEND` is set to `django.core.mail.backends.smtp.EmailBackend`
  - `SHIPPER_EMAIL_HOST_PASSWORD`
    - Email password
    - Required if `SHIPPER_EMAIL_BACKEND` is set to `django.core.mail.backends.smtp.EmailBackend`
  - `SHIPPER_EMAIL_USE_TLS`:
    - `0` -> disabled
    - `1` -> enabled (recommended for production)
    - Default: `1`
  - `SHIPPER_DEFAULT_FROM_EMAIL`
    - "From" field in emails, like "Joe McJoeFace \<joe@example.com\>"
    - Required if `SHIPPER_EMAIL_BACKEND` is set to `django.core.mail.backends.smtp.EmailBackend`
  - `SHIPPER_ADMIN_EMAILS`
    - A list of admins and their emails, delimited with : and ;
    - Format: `AdminName:adminemail@example.com;Admin2Name:admin2email@example.com`
    - Default: ``, disabled

### Backup
  - `SHIPPER_DBBACKUP_DIRECTORY`
    - Directory to back up databases and media files to when running `django-dbbackup` commands
    - The directory must exist beforehand; use `mkdir -p <full path to directory>` to achieve this.
    - Default: `/tmp/shipper-backup/`
  

### Sentry (error reporting):
  - `SHIPPER_SENTRY_SDK_DSN`
    - [Sentry](https://sentry.io) bug tracking
    - Recommended: https://34fd99861ec84ad2bd731c50267dc5f6@o444286.ingest.sentry.io/5418995 -> will help with future development of shipper
    - Default: ``, disabled
  - `SHIPPER_SENTRY_SDK_PII`
    - Send potential personally-identifiable information (PII) to Sentry when submitting bug reports
    - `0` -> disabled
    - Any other value -> enabled
    - Default: `0`
    - Warning: if enabled, shipper may send PII such as usernames and email addresses. Set to `False` in production.

## `.env.db`

  - `POSTGRES_USER`
    - Postgres user
    - Should match the `SHIPPER_SQL_USER` config key above.
  - `POSTGRES_PASSWORD`
    - Postgres user password
    - Should match the `SHIPPER_SQL_PASSWORD` config key above.
  - `POSTGRES_DB`
    - Postgres database name
    - Should match the `SHIPPER_SQL_DATABASE` config key above.


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog][keep-a-changelog].

[keep-a-changelog]: https://keepachangelog.com/en/1.0.0/

# [Unreleased]

[Unreleased]: https://github.com/shipperstack/shipper/compare/2.41.4...HEAD


# [2.41.4] - 2024-08-25

## Changed

- Fixed Docker image build crashing due to outdated Alpine Python package versions
- The task time limit value is now configurable with the `.env` file
- Updated dependencies (server)
  - django-constance[database] (3.1.0 -> 4.0.0)
  - django-dbbackup (4.1.0 -> 4.2.0)

[2.41.4]: https://github.com/shipperstack/shipper/compare/2.41.3...2.41.4


# [2.41.3] - 2024-08-23

## Added

- Added device note field to general API

## Changed

- Updated dependencies (shippy)
	- setuptools (72.2.0 -> 73.0.1)
- Updated dependencies (server)
	- @types/react (18.3.3 -> 18.3.4)
	- webpack (5.93.0 -> 5.94.0)
	- @babel/preset-env (7.25.3 -> 7.25.4)
	- django-celery-beat (2.6.0 -> 2.7.0)
	- django (5.0.8 -> 5.1)

[2.41.3]: https://github.com/shipperstack/shipper/compare/2.41.2...2.41.3


# [2.41.2] - 2024-08-18

## Added

- Added main internal admin page (for linking)

## Changed

- The internal admin page for tracking build mirror status now shows more stats for each item
- Fixed the nginx proxy configuration for layered reverse proxies (#562)
- Fixed the release helper ignoring Markdown headings during release (#563)
- General code cleanup

[2.41.2]: https://github.com/shipperstack/shipper/compare/2.41.1...2.41.2


# [2.41.1] - 2024-08-17

## Changed

- Fixed server image build failing due to outdated Alpine versions

[2.41.1]: https://github.com/shipperstack/shipper/compare/2.41.0...2.41.1

# [2.41.0] - 2024-08-17

This release contains a database migration.

## Added

- Added a feature model for Build objects (database only for now) (#354)
- Added checks to disallow Postgres-specific fields
- Added management commands for initializing a demo environment (#218)
- Added a device note edit screen for maintainers (#527)

## Changed
- Fixed some models not being picked up in the audit log
- Fixed a React key issue in a list on the frontend
- The SQLite3 backend correctly utilizes the given database name in the config
- Fixed issues loading the `crispy` form filters
- Clarified account management texts
- Fixed maintainer URLs as they were incorrectly being cached
- Fixed a bug with HTTP/HTTPS URL generation with the protocol forwarded by nginx (#562)
- General code cleanup
- Updated dependencies (server)
	- @babel/core (7.24.8 -> 7.25.2)
	- prettier (3.3.2 -> 3.3.3)
	- sentry-sdk (2.9.0 -> 2.13.0)
	- django-crispy-forms (2.2 -> 2.3)
	- typescript (5.5.3 -> 5.5.4)
	- @babel/preset-env (7.24.8 -> 7.25.3)
	- django (5.0.7 -> 5.0.8)
	- pyyaml (6.0.1 -> 6.0.2)
	- gunicorn (22.0.0 -> 23.0.0)
	- paramiko (3.4.0 -> 3.4.1)
- Updated dependencies (release-helper)
	- clap (4.5.9 -> 4.5.16)
	- regex (1.10.5 -> 1.10.6)
- Updated dependencies (shippy)
	- sentry-sdk (2.9.0 -> 2.13.0)
	- setuptools (70.3.0 -> 72.2.0)

[2.41.0]: https://github.com/shipperstack/shipper/compare/2.40.0...2.41.0

# [2.40.0] - 2024-07-13

## Changed

- The API now returns a one-click statistics and mirror-redirect download URL (#436)
- Fixed the server crashing on Windows due to a platform-dependent epoch conversion function call

## Removed

- Removed OpenAPI schema generation from shipper

[2.40.0]: https://github.com/shipperstack/shipper/compare/2.39.0...2.40.0

# [2.39.0] - 2024-07-13

This release contains a security vulnerability patch.

## Changed

- Updated dependencies (server)
  - sentry-sdk (2.5.0 -> 2.9.0)
  - prettier (3.3.1 -> 3.3.2)
  - webpack (5.91.0 -> 5.93.0)
  - django-crispy-forms (2.1 -> 2.2)
  - djangorestframework (3.15.1 -> 3.15.2)
  - humanize (4.9.0 -> 4.10.0)
  - django (5.0.6 -> 5.0.7)
  - @babel/core (7.24.7 -> 7.24.8)
  - @babel/preset-env (7.24.7 -> 7.24.8)
- Updated dependencies (release-helper)
  - clap (4.5.6 -> 4.5.9)
  - regex (1.10.4 -> 1.10.5)
  - git2 (0.18.3 -> 0.19.0)
- Updated dependencies (shippy)
  - sentry-sdk (2.5.0 -> 2.9.0)
  - setuptools (70.0.0 -> 70.3.0)
  - humanize (4.9.0 -> 4.10.0)
- The release-helper tool now includes Dependabot commits when generating the changelog
- Fixed the release-helper tool crashing when parsing non-standard dependency commits
- General code cleanup

[2.39.0]: https://github.com/shipperstack/shipper/compare/2.38.2...2.39.0

# [2.38.2] - 2024-06-08

## Added

- Added checkbox to unfilter disabled devices (#329)

## Changed

- Disabled devices no longer show via strikethrough text (#329)
- Disabled devices now show the color image on hover
- The device list now shows a message when the list is empty (#477)
- General code cleanup and refactoring

[2.38.2]: https://github.com/shipperstack/shipper/compare/2.38.1...2.38.2

# [2.38.1] - 2024-06-08

## Added

- The release helper now supports command aliases

## Changed

- Fixed faulty healthcheck for the nginx container in the Docker Compose configuration
- Fixed the release helper not checking the commit message length
- General code cleanup
- The release helper now sorts and organizes commits when writing the new changelog
- Updated dependencies (server)
  - requests (2.32.2 -> 2.32.3)
  - sentry-sdk (2.3.1 -> 2.5.0)
- Updated dependencies (release-helper)
  - clap (4.5.4 -> 4.5.6)
- Updated dependencies (shippy)
  - requests (2.32.2 -> 2.32.3)
  - sentry-sdk (2.3.1 -> 2.5.0)

[2.38.1]: https://github.com/shipperstack/shipper/compare/2.38.0...2.38.1

# [2.38.0] - 2024-05-27

This release contains a database migration.

## Added

- Added a notes field to the Device model

## Changed

- Fixed the server showing the "last updated" text on the device page when no builds are present
- Fixed the `x86_type` field preventing saving of non-x86 devices

server/config/settings.py
import os
import sentry_sdk
from django.core.exceptions import ImproperlyConfigured
from django.utils.translation import gettext_lazy
from kombu import Queue, Exchange
from paramiko import AuthenticationException, SSHException
from billiard.exceptions import TimeLimitExceeded
from sentry_sdk.integrations.django import DjangoIntegration
from pathlib import Path
from dotenv import load_dotenv

from .filters import IgnoreMissingBuild503Errors
from core.exceptions import UploadException, BuildMirrorException

load_dotenv()

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

SECRET_KEY = os.environ.get("SHIPPER_SECRET_KEY")
DEBUG = int(os.environ.get("SHIPPER_DEBUG", default=0))
try:
    ALLOWED_HOSTS = os.environ.get("SHIPPER_ALLOWED_HOSTS").split(" ")
except AttributeError:
    # Check will catch this error
    pass

try:
    CSRF_TRUSTED_ORIGINS = os.environ.get("SHIPPER_CSRF_TRUSTED_ORIGINS").split(" ")
except AttributeError:
    # Check will catch this error
    pass

CSRF_COOKIE_SECURE = int(os.environ.get("SHIPPER_CSRF_COOKIE_SECURE", default=1))
SESSION_COOKIE_SECURE = int(os.environ.get("SHIPPER_SESSION_COOKIE_SECURE", default=1))

SECURE_HSTS_SECONDS = int(os.environ.get("SHIPPER_SECURE_HSTS_SECONDS", default=0))

SECURE_PROXY_SSL_HEADER = ("HTTP_X_FORWARDED_PROTO", "https")


with open("version.txt") as v_file:
    SHIPPER_VERSION = v_file.readline().rstrip()

with open("shippy_compat_version.txt") as v_file:
    SHIPPER_SHIPPY_COMPAT_VERSION = v_file.readline().rstrip()


# Application definition
INSTALLED_APPS = [
    "constance",
    "django.contrib.admin",
    "django.contrib.auth",
    "accounts",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "rest_framework",
    "rest_framework.authtoken",
    "core",
    "downloads",
    "maintainers",
    "api",
    "internaladmin",
    "drf_chunked_upload",
    "auditlog",
    "django_celery_beat",
    "django_celery_results",
    "crispy_forms",
    "crispy_bootstrap5",
    "dbbackup",
    "django_cleanup.apps.CleanupConfig",  # must be last for successful deletions
]

MIDDLEWARE = [
    "django.middleware.cache.UpdateCacheMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.locale.LocaleMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "auditlog.middleware.AuditlogMiddleware",
    "config.middleware.SetIPMiddleware",
    "django.middleware.cache.FetchFromCacheMiddleware",
]

ROOT_URLCONF = "config.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR / "templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
                "config.context_processors.download_page_processor",
                "config.context_processors.version_processor",
                "config.context_processors.debug_mode_processor",
            ],
        },
    },
]

WSGI_APPLICATION = "config.wsgi.application"


# Database
# https://docs.djangoproject.com/en/3.0/ref/settings/#databases
SQL_ENGINE = os.environ.get("SHIPPER_SQL_ENGINE", default="django.db.backends.sqlite3")
SQL_DB_NAME = os.environ.get("SHIPPER_SQL_DATABASE")
if SQL_ENGINE == "django.db.backends.sqlite3":
    SQL_DB_NAME = str(BASE_DIR / f"{SQL_DB_NAME}.sqlite3")
DATABASES = {
    "default": {
        "ENGINE": SQL_ENGINE,
        "NAME": SQL_DB_NAME,
        "USER": os.environ.get("SHIPPER_SQL_USER", default="user"),
        "PASSWORD": os.environ.get("SHIPPER_SQL_PASSWORD", default="password"),
        "HOST": os.environ.get("SHIPPER_SQL_HOST", default="localhost"),
        "PORT": os.environ.get("SHIPPER_SQL_PORT", default="5432"),
    }
}

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

# Custom user model
AUTH_USER_MODEL = "accounts.User"

# Password validation
# https://docs.djangoproject.com/en/3.0/ref/settings/#auth-password-validators
AUTH_PASSWORD_VALIDATORS = [
    {"NAME": f"django.contrib.auth.password_validation.{name}"}
    for name in [
        "UserAttributeSimilarityValidator",
        "MinimumLengthValidator",
        "CommonPasswordValidator",
        "NumericPasswordValidator",
    ]
]


# Cache
CACHES = {
    "default": {
        "BACKEND": os.environ.get(
            "SHIPPER_CACHE_BACKEND",
            default="django.core.cache.backends.locmem.LocMemCache",
        ),
        "LOCATION": os.environ.get("SHIPPER_CACHE_LOCATION", default="shipper-cache"),
    }
}
CACHE_MIDDLEWARE_ALIAS = "default"
CACHE_MIDDLEWARE_SECONDS = 30
CACHE_MIDDLEWARE_KEY_PREFIX = "shipper"

# Logging
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "filters": {
        "ignore_missing_build_503_errors": {
            "()": IgnoreMissingBuild503Errors,
        },
        "require_debug_false": {
            "()": "django.utils.log.RequireDebugFalse",
        },
        "require_debug_true": {
            "()": "django.utils.log.RequireDebugTrue",
        },
    },
    "formatters": {
        "django.server": {
            "()": "django.utils.log.ServerFormatter",
            "format": "[{server_time}] {message}",
            "style": "{",
        }
    },
    "handlers": {
        "console": {
            "level": "INFO",
            "filters": ["require_debug_true", "ignore_missing_build_503_errors"],
            "class": "logging.StreamHandler",
        },
        "django.server": {
            "level": "INFO",
            "filters": ["ignore_missing_build_503_errors"],
            "class": "logging.StreamHandler",
            "formatter": "django.server",
        },
        "mail_admins": {
            "level": "ERROR",
            "filters": ["require_debug_false", "ignore_missing_build_503_errors"],
            "class": "django.utils.log.AdminEmailHandler",
        },
    },
    "loggers": {
        "django": {
            "handlers": ["console", "mail_admins"],
            "level": "INFO",
        },
        "django.server": {
            "handlers": ["django.server"],
            "level": "INFO",
            "propagate": False,
        },
    },
}


# Email
EMAIL_BACKEND = os.environ.get(
    "SHIPPER_EMAIL_BACKEND", default="django.core.mail.backends.console.EmailBackend"
)
EMAIL_HOST = os.environ.get("SHIPPER_EMAIL_HOST", default="")
EMAIL_PORT = os.environ.get("SHIPPER_EMAIL_PORT", default="")
EMAIL_HOST_USER = os.environ.get("SHIPPER_EMAIL_HOST_USER", default="")
EMAIL_HOST_PASSWORD = os.environ.get("SHIPPER_EMAIL_HOST_PASSWORD", default="")
EMAIL_USE_TLS = int(os.environ.get("SHIPPER_EMAIL_USE_TLS", default=1))
DEFAULT_FROM_EMAIL = os.environ.get("SHIPPER_DEFAULT_FROM_EMAIL", default="")

ADMINS = [
    tuple(i.split(":"))
    for i in os.environ.get("SHIPPER_ADMIN_EMAILS", default="").split(";")
]
MANAGERS = ADMINS

# Internationalization
# https://docs.djangoproject.com/en/3.0/topics/i18n/
LANGUAGE_CODE = "en-us"
TIME_ZONE = "UTC"
USE_I18N = True
USE_TZ = True

LANGUAGES = [
    ("ko", gettext_lazy("Korean")),
    ("en", gettext_lazy("English")),
]

LOCALE_PATHS = (BASE_DIR / "locale",)

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.0/howto/static-files/
STATIC_URL = "/static/"
STATIC_ROOT = BASE_DIR / "static"

# Media
MEDIA_URL = "/media/"
MEDIA_ROOT = BASE_DIR / "media"

# Login
LOGIN_REDIRECT_URL = "/maintainers/"

# Django REST Framework
REST_FRAMEWORK = {
    "DEFAULT_RENDERER_CLASSES": ("rest_framework.renderers.JSONRenderer",),
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "rest_framework.authentication.TokenAuthentication",
    ),
    "DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.IsAuthenticated",),
    "DEFAULT_THROTTLE_CLASSES": [
        "rest_framework.throttling.AnonRateThrottle",
        "rest_framework.throttling.UserRateThrottle",
    ],
    "DEFAULT_THROTTLE_RATES": {"anon": "10/minute", "user": "10/second"},
}


# drf-chunked-upload
DRF_CHUNKED_UPLOAD_COMPLETE_EXT = ""
DRF_CHUNKED_UPLOAD_ABSTRACT_MODEL = False
DRF_CHUNKED_UPLOAD_CHECKSUM = os.environ.get(
    "SHIPPER_UPLOAD_CHECKSUM", default="sha256"
)
DRF_CHUNKED_UPLOAD_MAX_BYTES = 5_000_000_000  # 5 GB


# Celery
CELERY_BROKER_URL = "pyamqp://rabbitmq:5672/"
CELERY_TASK_TIME_LIMIT = 60 * int(os.environ.get("SHIPPER_TASK_TIME_LIMIT", default=60))
CELERY_TASK_TRACK_STARTED = True
CELERY_RESULT_BACKEND = "django-db"
CELERY_RESULT_CACHE = "default"
CELERY_TASK_SERIALIZER = "json"
CELERY_RESULT_SERIALIZER = "json"
CELERY_ACCEPT_CONTENT = ["application/json"]
CELERY_RESULT_EXTENDED = True
CELERY_DEFAULT_QUEUE = "default"
CELERY_QUEUES = (
    Queue("default", Exchange("default"), routing_key="default"),
    Queue("mirror_upload", Exchange("mirror_upload"), routing_key="mirror_upload"),
)


# Constance
CONSTANCE_BACKEND = "constance.backends.database.DatabaseBackend"
CONSTANCE_CONFIG = {
    "SHIPPER_MAIN_WEBSITE_URL": (
        "#",
        "Link to redirect users to when they click on the 'Back to main website' "
        "button on the top of the page. Setting this configuration to the pound "
        "symbol disables redirection and hides the button.",
        str,
    ),
    "SHIPPER_DOWNLOADS_PAGE_MAIN_BRANDING": (
        "Downloads",
        "Sets the navbar branding at the top left of the website.",
        str,
    ),
    "SHIPPER_DOWNLOADS_PAGE_DONATION_URL": (
        "#",
        "URL to refer users to when they wish to donate.",
        str,
    ),
    "SHIPPER_DOWNLOADS_PAGE_DONATION_MESSAGE": (
        "Please consider donating, thank you!",
        "Donation message to show to users in the donation banner.",
        str,
    ),
    "SHIPPER_DOWNLOADS_DISABLE_MAIN_SERVER": (
        False,
        "Disables downloads from the main server. Enable only if you are experiencing "
        "load problems.",
        bool,
    ),
    "SHIPPER_DOWNLOADS_ARCHIVE_THROTTLE": (
        1,
        "Throttles the download speed of archived builds, in MB/s.",
        int,
    ),
    "SHIPPER_FILE_NAME_FORMAT": (
        "[A-Za-z]+-(?P<version>[a-z0-9.]+)-(?P<codename>[A-Za-z0-9_]+)-OFFICIAL-("
        "?P<variant>[a-z]+)-(?P<date>[0-9]{8}).zip",
        "Regex pattern to use when parsing file names of uploaded artifacts. The "
        "pattern must include the following four named match groups; otherwise an "
        "exception will occur during uploading: `version', `codename`, `variant`, "
        "and `date`.",
        str,
    ),
    "SHIPPER_ALLOWED_VERSIONS_TO_UPLOAD": (
        "*",
        "Versions that maintainers can currently upload to the system. Wildcards are "
        "supported. Specify multiple versions on each line.",
        str,
    ),
    "SHIPPER_BUILD_ARCHIVE_DAYS": (
        90,
        "Builds that are 'archived' in the system after this many number of days. "
        "Archived builds are not mirrored to any mirror servers.",
        int,
    ),
    "SHIPPER_ENABLE_MIRRORING": (
        True,
        "Enables the mirroring functionality of shipper. Disable temporarily if you "
        "are experiencing problems.",
        bool,
    ),
}
CONSTANCE_CONFIG_FIELDSETS = {
    "Downloads page": (
        "SHIPPER_MAIN_WEBSITE_URL",
        "SHIPPER_DOWNLOADS_PAGE_MAIN_BRANDING",
        "SHIPPER_DOWNLOADS_PAGE_DONATION_URL",
        "SHIPPER_DOWNLOADS_PAGE_DONATION_MESSAGE",
    ),
    "Download": (
        "SHIPPER_DOWNLOADS_DISABLE_MAIN_SERVER",
        "SHIPPER_DOWNLOADS_ARCHIVE_THROTTLE",
    ),
    "Upload": (
        "SHIPPER_FILE_NAME_FORMAT",
        "SHIPPER_ALLOWED_VERSIONS_TO_UPLOAD",
    ),
    "Maintenance": ("SHIPPER_BUILD_ARCHIVE_DAYS",),
    "Mirroring": ("SHIPPER_ENABLE_MIRRORING",),
}

# django-auditlog
AUDITLOG_INCLUDE_TRACKING_MODELS = ("constance.Constance",)

# Django Crispy Forms
CRISPY_ALLOWED_TEMPLATE_PACKS = "bootstrap5"
CRISPY_TEMPLATE_PACK = "bootstrap5"


# Django-dbbackup
DBBACKUP_STORAGE = "django.core.files.storage.FileSystemStorage"
DBBACKUP_STORAGE_OPTIONS = {
    "location": os.environ.get(
        "SHIPPER_DBBACKUP_DIRECTORY", default="/tmp/shipper-backup/"
    )
}

# django-ipware
IPWARE_META_PRECEDENCE_ORDER = (
    "HTTP_CF_CONNECTING_IP",
    "HTTP_X_FORWARDED_FOR",
    "X_FORWARDED_FOR",
    "HTTP_CLIENT_IP",
    "HTTP_X_REAL_IP",
    "HTTP_X_FORWARDED",
    "HTTP_X_CLUSTER_CLIENT_IP",
    "HTTP_FORWARDED_FOR",
    "HTTP_FORWARDED",
    "HTTP_VIA",
    "REMOTE_ADDR",
)


# Sentry SDK
def before_send(event, hint):
    if "log_record" in hint:
        if hint["log_record"].name == "django.security.DisallowedHost":
            return None
    return event


if DEBUG == 1:
    sentry_transaction_rate = 1.0
else:
    sentry_transaction_rate = 0.2


sentry_sdk.init(
    dsn=os.environ.get("SHIPPER_SENTRY_SDK_DSN", default=""),
    integrations=[DjangoIntegration()],
    release=f"{SHIPPER_VERSION}",
    traces_sample_rate=sentry_transaction_rate,
    send_default_pii=(int(os.environ.get("SHIPPER_SENTRY_SDK_PII", default=0)) == 0),
    before_send=before_send,
    ignore_errors=[
        SSHException,
        AuthenticationException,
        ConnectionRefusedError,
        TimeLimitExceeded,
        KeyboardInterrupt,
        UploadException,
        BuildMirrorException,
        ImproperlyConfigured,
    ],
)
